File: agent/src/reporting_agent/events.py
"""The SSE event vocabulary — one contract, expressed in two languages.

The declaration below is mirrored in `app/lib/events.ts`, and a static guard in the
web test suite (`app/test/event-mirror.static.test.ts`) extracts the quoted strings
from between the sentinel comments in both files and compares the two sets
(Req 40.13). That is why the literals sit on their own, between sentinels, rather
than inside an `Enum` or a class: the guard needs neither a Python parser nor a
TypeScript parser, so the guard itself cannot drift from what it guards.

**The full vocabulary is declared; a subset is emitted.** This spec drives six of the
ten types. `delta`, `chart`, `verification` and `report_file` belong to the specs that
add prose, charts and the compile/render/verify pipeline — those specs add *emitters*,
not vocabulary, so the mirror never has to be renegotiated (Req 14.11 with 40.13), and
a client that meets an unhandled type ignores it and keeps reading (Req 40.6).

Nothing here emits. `main.py` owns the single egress function every event passes
through, and `EVENT_TYPES` is what it validates a `type` field against (Req 14.15).
"""

from typing import Final

# --- BEGIN EVENT TYPES (mirrored in app/lib/events.ts) ---
EVENT_TYPES: Final[tuple[str, ...]] = (
    "delta",
    "tool",
    "progress",
    "heartbeat",
    "snapshot_ready",
    "chart",
    "verification",
    "report_file",
    "error",
    "done",
)
# --- END EVENT TYPES ---

# The six types this spec's runtime actually emits (Req 14.11). Deliberately outside
# the sentinels: the mirror guard reads every quoted string between them, so a second
# list of type names inside would fail the comparison against the TypeScript side.
#
# `verification` and `report_file` are absent on purpose — no document is produced
# here, so neither may ever be emitted, and the ordering guarantee that a
# `report_file` never arrives without a passing `verification` before it cannot be
# violated by a runtime that emits neither.
EMITTED_BY_FOUNDATION: Final[frozenset[str]] = frozenset(
    {"tool", "progress", "heartbeat", "snapshot_ready", "error", "done"}
)

# The terminal event, and the only one permitted to be last (Req 14.10).
TERMINAL_EVENT_TYPE: Final[str] = "done"

# The keep-alive type, named here so `heartbeat.py` imports it instead of re-declaring
# the literal — one spelling of the string, in the module that owns the vocabulary.
# Deliberately outside the sentinels for the same reason TERMINAL_EVENT_TYPE is: the
# mirror guard reads *every* quoted string between them, so a second occurrence of
# "heartbeat" inside would fail the comparison against the TypeScript side.
HEARTBEAT_EVENT_TYPE: Final[str] = "heartbeat"

# The two `tool` step names this spec drives (Req 14.7). Declared here rather than in
# `main.py` because two modules consume them — `main.py`, which owns the step vocabulary
# the timeline renders, and `collect/pipeline.py`, which opens the steps — and a step name
# spelled twice is a step the UI renders as two different things. Outside the sentinels:
# these are not event *types*, so the TypeScript mirror does not carry them.
TOOL_COLLECT_INVENTORY: Final[str] = "collect_inventory"
TOOL_COLLECT_METRICS: Final[str] = "collect_metrics"

# The four the document phases drive. Same reasoning: two modules consume each name —
# `report_pipeline.py`, which opens the step, and the timeline that renders it — and a step
# name spelled twice is a step the UI renders as two different things.
TOOL_COMPILE_FIGURES: Final[str] = "compile_figures"
TOOL_RENDER_DOCUMENT: Final[str] = "render_document"
TOOL_VERIFY_DOCUMENT: Final[str] = "verify_document"
TOOL_UPLOAD_ARTIFACT: Final[str] = "upload_artifact"

# Every type the **document** pipeline emits (Req 42.1). All ten, because this spec adds the
# four the foundation deliberately left unemitted — and it adds emitters, not vocabulary, so
# `EVENT_TYPES` and `app/lib/events.ts` are untouched and the cross-language mirror guard
# never has to be renegotiated. Outside the sentinels, like every other non-vocabulary name
# here: the guard reads every quoted string between them.
#
# **Named rather than derived from `EVENT_TYPES`.** `frozenset(EVENT_TYPES)` would make the
# equality assertion below a tautology, and an eleventh vocabulary type would then be
# silently claimed as emitted by a pipeline that has no emitter for it. Spelled out, the
# assertion is the question "did you write the emitter?" asked at import.
EMITTED_BY_REPORT_PIPELINE: Final[frozenset[str]] = frozenset(
    {
        "delta",
        "tool",
        "progress",
        "heartbeat",
        "snapshot_ready",
        "chart",
        "verification",
        "report_file",
        "error",
        "done",
    }
)

PROGRESS_UNIT_BLOCKS: Final[str] = "blocks"


def is_declared_event_type(value: object) -> bool:
    """Is `value` one of the declared event types (Req 14.15)?"""
    return isinstance(value, str) and value in EVENT_TYPES


# A declaration that contradicts itself is worth catching at import rather than at the
# first emission: EMITTED_BY_FOUNDATION is a subset of the vocabulary, and no type is
# declared twice.
assert EMITTED_BY_FOUNDATION <= frozenset(EVENT_TYPES), EMITTED_BY_FOUNDATION
assert len(set(EVENT_TYPES)) == len(EVENT_TYPES), EVENT_TYPES
assert TERMINAL_EVENT_TYPE in EMITTED_BY_FOUNDATION, TERMINAL_EVENT_TYPE
assert HEARTBEAT_EVENT_TYPE in EMITTED_BY_FOUNDATION, HEARTBEAT_EVENT_TYPE
assert HEARTBEAT_EVENT_TYPE != TERMINAL_EVENT_TYPE
assert EMITTED_BY_FOUNDATION < EMITTED_BY_REPORT_PIPELINE, EMITTED_BY_REPORT_PIPELINE
assert EMITTED_BY_REPORT_PIPELINE == frozenset(EVENT_TYPES), EMITTED_BY_REPORT_PIPELINE

File: app/lib/events.ts
/**
 * The SSE event vocabulary — one contract, expressed in two languages.
 *
 * The declaration below is mirrored in `agent/src/reporting_agent/events.py`
 * (Requirement 40.7), and `app/test/event-mirror.static.test.ts` extracts the
 * quoted strings from between the sentinel comments in both files and compares
 * the two sets (Requirement 40.13). The literals therefore sit on their own,
 * between sentinels, rather than inside a mapped type or an enum: the guard
 * needs neither a TypeScript parser nor a Python parser, so the guard itself
 * cannot drift from what it guards.
 *
 * **The full vocabulary is declared; a subset is emitted.** This spec's runtime
 * drives six of the ten types. `delta`, `chart`, `verification` and
 * `report_file` belong to the specs that add prose, charts and the
 * compile/render/verify pipeline — those specs add *emitters*, not vocabulary,
 * so the mirror never has to be renegotiated.
 *
 * Pure, secret-free and shared with the client, so deliberately not
 * `server-only`: the stream-parsing hook imports `isDeclaredEventType` to
 * ignore an event type it does not know rather than crash on it
 * (Requirement 40.6).
 */

// --- BEGIN EVENT TYPES (mirrored in agent/src/reporting_agent/events.py) ---
export const EVENT_TYPES = [
  "delta",
  "tool",
  "progress",
  "heartbeat",
  "snapshot_ready",
  "chart",
  "verification",
  "report_file",
  "error",
  "done",
] as const
// --- END EVENT TYPES ---

/** One of the ten declared event types. */
export type EventType = (typeof EVENT_TYPES)[number]

/** The terminal event, and the only one permitted to be last. */
export const TERMINAL_EVENT_TYPE = "done" satisfies EventType

const DECLARED: ReadonlySet<string> = new Set<string>(EVENT_TYPES)

/**
 * Is this a declared event type?
 *
 * An older client meeting a newer type must degrade, not crash: an undeclared
 * type is ignored, applies no state change, and the stream keeps being read
 * (Requirement 40.6). That is a narrowing predicate rather than a throwing
 * parse for exactly that reason.
 */
export function isDeclaredEventType(value: unknown): value is EventType {
  return typeof value === "string" && DECLARED.has(value)
}

// --- Per-phase constants ----------------------------------------------------

/**
 * The unit a `progress` event's counts are in, **per phase** (Requirement 40.15).
 *
 * Declared here as a constant rather than carried on the run row, and the
 * requirement is explicit about that: `unit` comes "from the unit that phase
 * declares in `app/lib/events.ts` as a per-phase constant rather than as run
 * state". So there is no `progress_unit` column, and the relay cannot emit a unit
 * a client has never seen — the vocabulary owns it.
 *
 * Keyed by the run status the phase corresponds to, because that is what the relay
 * has in hand: Requirement 40.15 also sets a `progress` event's `id` from the row's
 * `status`, so one lookup serves both fields and they cannot disagree about which
 * phase is being described.
 *
 * `collecting` is the only phase this spec drives that carries a countable unit of
 * work. The three undriven phases are absent rather than mapped to a placeholder: a
 * status with no entry here is a status that emits no `progress` event, which is the
 * same outcome Requirement 40.14 requires of a phase whose counts are absent.
 */
export const PHASE_PROGRESS_UNIT: Readonly<Record<string, string>> =
  Object.freeze({
    collecting: "resources",
  })

/**
 * The `tool` step this spec's relay opens for each non-terminal phase.
 *
 * `name` is drawn from the tool names the agent's own `tool` events use, so a client
 * that later receives real agent events and relay-derived ones renders one timeline
 * rather than two. `label` is the badge the activity timeline shows.
 *
 * The relay derives these from the row's `status`, which is why the `id` of a
 * `progress` event is that same status: the step and the progress bar attached to it
 * are the same phase, named the same way, with no correlation table in between.
 */
export const PHASE_TOOL_STEP: Readonly<
  Record<string, { readonly name: string; readonly label: string }>
> = Object.freeze({
  queued: Object.freeze({ name: "collect_inventory", label: "Queued" }),
  claimed: Object.freeze({ name: "collect_inventory", label: "Starting" }),
  collecting: Object.freeze({ name: "collect_metrics", label: "Collecting" }),
})

// --- Relay timings ----------------------------------------------------------

/**
 * How often the relay re-reads the `report_runs` row (Requirement 40.10).
 *
 * Two seconds, which is what makes the determinate bar at most ~7 seconds stale
 * worst case — the agent's 5-second progress throttle plus this poll — on a run that
 * lasts 8 to 12 minutes. Shared with the hook so the client's reconnect budget is
 * expressed against the same number the server polls at.
 */
export const RELAY_POLL_MS = 2_000

/** How often a `heartbeat` is emitted while the row is non-terminal (Req 40.10). */
export const RELAY_HEARTBEAT_MS = 15_000

/**
 * How long the relay will emit nothing but heartbeats before closing
 * (Requirement 40.3).
 *
 * The relay is a **live view whose loss costs nothing**, so closing it is cheaper
 * than keeping a connection alive through an intermediary that may cut it anyway —
 * CloudFront defaults to a 30-second origin-response timeout and an ALB to a
 * 60-second idle timeout, and both kill SSE far more often than any app-level limit
 * does. For a run sitting in `collecting` for ten minutes this means the relay closes
 * roughly every two minutes and the client reopens; that churn is intentional,
 * because a disposable view that reconnects cleanly is strictly better than a
 * long-lived one that has to be correct.
 */
export const RELAY_IDLE_CLOSE_MS = 120_000

/**
 * How long the client waits before reopening a stream for a non-terminal run
 * (Requirement 40.11).
 *
 * Under the 5 seconds the requirement allows, and deliberately not zero: an
 * immediate reopen against a relay that is closing for a reason other than its idle
 * window — a deploy roll, a proxy — turns one dropped connection into a reconnect
 * loop.
 */
export const RELAY_REOPEN_MS = 2_000
